### PR TITLE
feat(cli/templates) add clojurescript

### PR DESCRIPTION
**Pull request by just-sultanov**
### What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?

- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [x] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/create-tauri-app/blob/dev/.changes/readme.md).
- [x] I have added a convincing reason for adding this feature, if necessary

### Other information

Related to this nuxt/framework#184

### CODE REVIEW

File: packages/cli/fragments/fragment-clojurescript/_cta_manifest_ [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+# SPDX-License-Identifier: MIT
+
+beforeDevCommand = {{pkg_manager_run_command}} dev
+beforeBuildCommand = {{pkg_manager_run_command}} build
+devPath = http://localhost:3000
+distDir = ../dist
+withGlobalTauri = true
+
+[files]
+cljs.svg = public/cljs.svg
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-clojurescript/_cta_manifest_]: remove this

File: packages/cli/fragments/_assets_/cljs.svg [hunk 2]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (amrbashir) [inline packages/cli/fragments/_assets_/cljs.svg]: move this file to `fragment-clojurescript/public`

File: packages/cli/fragments/fragment-clojurescript/shadow-cljs.edn [hunk 3]
```diff
@@ -0,0 +1,23 @@
+{:deps {:aliases [:develop :test]}
+
+ :nrepl {:init-ns user}
+
+ :dev-http {3000 "public"
+            3001 "target/test"}
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-clojurescript/shadow-cljs.edn]:
```suggestion
 :dev-http {1420"public"
            1421 "target/test"}
```

File: packages/cli/fragments/fragment-clojurescript/_cta_manifest_ [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2019-2022 Tauri Programme within The Commons Conservancy
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-License-Identifier: MIT
+
+beforeDevCommand = {{pkg_manager_run_command}} dev
+beforeBuildCommand = {{pkg_manager_run_command}} build
+devPath = http://localhost:3000
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-clojurescript/_cta_manifest_]:
```suggestion
devPath = http://localhost:1420
```

File: README.md [hunk 5]
```diff
@@ -85,6 +85,7 @@ Currently supported template presets include:
 - `next-ts`
 - `preact`
 - `preact-ts`
+- `clojurescript`
```
Comment (amrbashir) [inline README.md:90]: you should do the same in `packages/cli/node/README.md`

File: .github/workflows/templates-test.yml [hunk 6]
```diff
@@ -115,6 +115,10 @@ jobs:
           sudo apt-get update
           sudo apt-get install -y libgtk-3-dev webkit2gtk-4.0 libayatana-appindicator3-dev librsvg2-dev patchelf
 
+      - run: |
+          sudo apt install -y clojure
+        if: matrix.settings.template == 'clojurescript'
+
```
Comment (amrbashir) [inline .github/workflows/templates-test.yml]: move this up before `download cli` step

File: packages/cli/fragments/fragment-clojurescript/src/main/clojure/app/core.cljs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  (* x x))
+
+
+(defn mount-root
+  "Mount root component."
+  {:dev/after-load true}
+  []
+  (some->>
+   (gdom/getElement "root")
+   (dom/render [:h1 "Welcome to Tauri!"])))
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-clojurescript/src/main/clojure/app/core.cljs]: This is a very very basic example, there should be an example showcasing how to invoke a rust command using https://tauri.app/v1/api/js/tauri#invoke

Also you need to match the same structure and functionality of the other templates which uses the css from `base/src/styles.css` and have a simple example for invoking rust command

**Review by amrbashir — CHANGES_REQUESTED** [review]
You also need to add a post init note about installing `java` and `clojure`
A similar thing has been done for `yew` template in `packages/cli/src/templates.rs` in `post_init_info` method

Also don't forget to run `carg fmt --all` and `pnpm format` in the repo root

Thanks.

File: packages/cli/fragments/fragment-clojurescript/src/test/clojure/app/core_test.cljs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+(ns app.core-test
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-clojurescript/src/test/clojure/app/core_test.cljs]: test related, remove

File: packages/cli/fragments/fragment-clojurescript/package.json [hunk 9]
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "{{package_name}}",
+  "private": true,
+  "version": "0.0.0",
+  "scripts": {
+    "assets:build": "cp ./src/style.css ./public/",
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-clojurescript/package.json]: why is this necessary? can't you import the style inside .cljs file directly?

File: packages/cli/fragments/fragment-clojurescript/package.json [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    "preshadow:watch": "npm run assets:build",
+    "shadow:watch": "npm run shadow-cljs watch app test",
+    "preshadow:build": "npm run assets:build",
+    "shadow:build": "npm run shadow-cljs release app",
+    "preshadow:test": "npm run assets:build",
+    "shadow:test": "npm run shadow-cljs compile ci && npm run karma start karma.config.js --single-run",
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-clojurescript/package.json]: test related, remove this along with anything relates to testing in this file, depdencies or scripts

File: packages/cli/fragments/fragment-clojurescript/karma.config.js [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    customLaunchers: {
+      ChromeHeadlessNoSandbox: {
+        base: "ChromeHeadless",
+        flags: ["--no-sandbox"],
+      },
+    },
+    basePath: "target",
+    files: ["test/js/ci.js"],
+    frameworks: ["cljs-test"],
+    plugins: ["karma-cljs-test", "karma-chrome-launcher"],
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-clojurescript/karma.config.js]: test related, remove

**Review by amrbashir — CHANGES_REQUESTED** [review]
I should've caught this in my review but the template shouldn't include testing so please remove all test related files/deps/scripts...etc

File: packages/cli/fragments/fragment-clojurescript/package.json (hunk 9 above)
Comment (just-sultanov) [reply packages/cli/fragments/fragment-clojurescript/package.json]: Unfortunately, we cannot import styles into cljs

Comment (amrbashir) [reply packages/cli/fragments/fragment-clojurescript/package.json]: then you need to move `base/src/style.css` to `_assets_/style.css` and add it to the `[files]` section in `_cta_manifest_` but you'd have to do the same for other fragments

Comment (just-sultanov) [reply packages/cli/fragments/fragment-clojurescript/package.json]: Oh, super! Thanks 👍

File: packages/cli/fragments/fragment-clojurescript/.vscode/extensions.json [hunk 12]
```diff
@@ -0,0 +1,3 @@
+{
+  "recommendations": ["tauri-apps.tauri-vscode", "rust-lang.rust-analyzer"]
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-clojurescript/.vscode/extensions.json]: Seems like clojurescript has a vscode extension, so we should recommend that
```suggestion
  "recommendations": ["tauri-apps.tauri-vscode", "rust-lang.rust-analyzer", "betterthantomorrow.calva"]
```

File: packages/cli/fragments/fragment-clojurescript/README.md [hunk 13]
```diff
@@ -0,0 +1,7 @@
+# Tauri + ClojureScript
+
+This template should help get you started developing with Tauri and ClojureScript + shadow-cljs.
+
+## Recommended IDE Setup
+
+- [VS Code](https://code.visualstudio.com/) + [Tauri](https://marketplace.visualstudio.com/items?itemName=tauri-apps.tauri-vscode) + [rust-analyzer](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust-analyzer)
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-clojurescript/README.md]:
```suggestion
- [VS Code](https://code.visualstudio.com/) + [Tauri](https://marketplace.visualstudio.com/items?itemName=tauri-apps.tauri-vscode) + [rust-analyzer](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust-analyzer) + [calva](https://marketplace.visualstudio.com/items?itemName=betterthantomorrow.calva)
```

File: packages/cli/fragments/fragment-clojurescript/deps.edn [hunk 14]
```diff
@@ -0,0 +1,9 @@
+{:paths ["src/main/clojure"]
+
+ :deps {reagent/reagent {:mvn/version "1.1.1"
+                         :exclusions [cljsjs/react
+                                      cljsjs/react-dom
+                                      cljsjs/react-dom-server]}}
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-clojurescript/deps.edn]: this package seems to be deprecated https://clojars.org/cljsjs/react-dom-server

File: packages/cli/fragments/fragment-clojurescript/package.json [hunk 15]
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "{{package_name}}",
+  "private": true,
+  "version": "0.0.0",
+  "scripts": {
+    "dev": "npm run shadow-cljs watch app",
+    "build": "npm run shadow-cljs release app",
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-clojurescript/package.json]:
```suggestion
    "dev": "shadow-cljs watch app",
    "build": "shadow-cljs release app",
```

File: packages/cli/fragments/fragment-clojurescript/src/main/clojure/app/core.cljs [hunk 16]
```diff
@@ -0,0 +1,54 @@
+(ns app.core
+  (:require
+    ["@tauri-apps/api/tauri" :as tauri]
+    [goog.dom :as gdom]
+    [reagent.core :as r]
+    [reagent.dom :as dom]))
+
+;;  Learn more about Tauri commands at https://tauri.app/v1/guides/features/command
```
Comment (amrbashir) [inline packages/cli/fragments/fragment-clojurescript/src/main/clojure/app/core.cljs]: Can you move this comment right above the `invoke` call?

**Review by amrbashir — CHANGES_REQUESTED** [review]

**Comment by amrbashir** [comment]
CI seems to be failing to build the clojure project

File: packages/cli/fragments/fragment-clojurescript/deps.edn (hunk 14 above)
Comment (just-sultanov) [reply packages/cli/fragments/fragment-clojurescript/deps.edn]: This record means that we want to exclude some dependencies and use `React` directly from `npm`. But I double-checked and this record unnecessary. I've removed this part.

**Comment by just-sultanov** [comment]
> CI seems to be failing to build the clojure project

I don't know why `deps.edn` can't be found. 
OK, I'll switch to using only `shadow-cljs` without `deps.edn`. Local testing of commands on `mac/ubuntu` passes without any problems.

**Comment by just-sultanov** [comment]
@amrbashir all tests were passed. Only `pnpm` doesn't work.

Cause:
`The required JS dependency "object-assign" is not available, it has been requested "node_modules/react/cjs/react.production.min.js".`

I can fix it using the following methods:
1) Add  `.npmrc` to the root of the clojurescript template to allow install dependencies using `pnpm i --shamefully-hoist`
2) Add `object-assign` dependency to `package.json`
3) Support only `yarn` and `npm` package managers

What do you think about it?

**Comment by nothingismagick** [comment]
In my opinion #2 seems the best, because #3 is the worst.

I read through similar troubles over at nuxt (maybe there is another solution?
https://github.com/nuxt/nuxt.js/issues/14146

**Comment by amrbashir** [comment]
yeah option no.2 seems to be the best for now

**Comment by just-sultanov** [comment]
Unfortunately, option no.2 didn't work. I added a `.npmrc` file with the required parameters for the specified dependencies (without wildcards). In addition, I have added comments describing the problem with the compiler.

**Comment by amrbashir** [comment]
That's fine, you can also create this file only if pnpm is the package manager so the file name would be like this `_[pnpm]_.npmrc`

**Comment by just-sultanov** [comment]
> That's fine, you can also create this file only if pnpm is the package manager so the file name would be like this `_[pnpm]_.npmrc`

done

**Review by amrbashir — DISMISSED** [review]

**Comment by amrbashir** [comment]
Thank you

File: .changes/clojurescript.md [hunk 17]
```diff
@@ -0,0 +1,6 @@
+---
+"create-tauri-app": patch
+"create-tauri-app-js": patch
```
Comment (amrbashir) [inline .changes/clojurescript.md]:
```suggestion
"create-tauri-app": minor
"create-tauri-app-js": minor
```

**Review by amrbashir — APPROVED** [review]

**Comment by phrohdoh** [comment]
It appears as if these changes were lost sometime between them being merged and the moment I'm writing this comment as the `.cljs` files, the `Template::ClojureScript` match arm, and the templated files are not on dev (bca3828ae1f10c476579b553213713d8e843908e).

Any idea what happened here?

**Comment by FabianLars** [comment]
Many templates were removed in https://github.com/tauri-apps/create-tauri-app/pull/284
https://tauri.app/blog/2023/03/01/create-tauri-app-version-3-released#removing-templates
